ChemistryLib/PhreeqcKernelData/CreateReactionRate.cpp

```diff
@@ -0,0 +1,72 @@
+/**
+ * \copyright
+ * Copyright (c) 2012-2019, OpenGeoSys Community (http://www.opengeosys.org)
+ *            Distributed under a Modified BSD License.
+ *              See accompanying file LICENSE.txt or
+ *              http://www.opengeosys.org/project/license
+ *
+ */
+
+#include <stdlib.h>
+#include <boost/optional/optional.hpp>
+
+#include "BaseLib/ConfigTree.h"
+#include "CreateReactionRate.h"
+
+#include "ThirdParty/iphreeqc/src/src/phreeqcpp/global_structures.h"
+
+namespace ChemistryLib
+{
+namespace PhreeqcKernelData
+{
+std::tuple<rate*, int> createReactionRates(
+    boost::optional<BaseLib::ConfigTree> const& config)
+{
+    if (!config)
+    {
+        return {};
+    }
+
+    int count_rates = 0;
+    struct rate* rates = (struct rate*)malloc(sizeof(struct rate));
+    for (auto const& rate_config :
+         //! \ogs_file_param{prj__chemical_system__rates__rate}
+         config->getConfigSubtreeList("rate"))
+    {
+        if (count_rates > 0)
+            rates = (struct rate*)realloc(
```

Comment:
direct memory allocation is highly discouraged.
`unique_ptr`, `vector`.

Reply:
partially solved.

Reply:
Memory reallocation appears once in the small private class method `void PhreeqcKernel::reinitializeRates()`.